### PR TITLE
Analyzer layer

### PR DESCRIPTION
**Pull request by Rythmag**
1. @webservlet("fulltext_query") to check the results. Creates the client connection and passes to FulltextSearchQuery along with name of index to be searched.
2. FulltextSearchQuery search for keywords in the in the index and returns hits as json object.
3. Keywords.java lists all the keywords used. For term matching OR logic is use i.e. A log line will be in the searchhits if it conatins one or more of the keywords.
4. The search is done on a temporary index file for now.

### CODE REVIEW

File: error-analyzer/src/main/java/com/google/sps/FulltextSearchQuery.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+import org.elasticsearch.search.SearchHit;
+import org.elasticsearch.search.SearchHits;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.client.RequestOptions;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class FulltextSearchQuery {
```
Comment (nishantnahata) [inline error-analyzer/src/main/java/com/google/sps/FulltextSearchQuery.java]: Please add javadoc and 1-2lines for public functions so it is better to understand. Thanks.
For other files as well.

File: src/main/java/com/google/error_analyzer/servlets/FulltextSearch.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+public class FulltextSearch extends HttpServlet {
+
+    private String indexFile = "trial_index"; //later fetched from request
+    private static final Logger logger = LogManager.getLogger(FulltextSearch.class);
+
+    @Override
+    public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+        response.setContentType("application/json");
+        try{
+            RestHighLevelClient client = new RestHighLevelClient(RestClient.builder(new HttpHost("35.194.181.238", 9200, "http")));
```
Comment (nishantnahata) [inline src/main/java/com/google/error_analyzer/servlets/FulltextSearch.java:47]: please remove the host ip

File: src/main/java/com/google/error_analyzer/servlets/FulltextSearch.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            FulltextSearchQuery searchQuery = new FulltextSearchQuery();
```
Comment (nishantnahata) [inline src/main/java/com/google/error_analyzer/servlets/FulltextSearch.java:48]: Make it a class variable.

File: src/main/java/com/google/error_analyzer/servlets/FulltextSearch.java (hunk 2 above)
Comment (nishantnahata) [inline src/main/java/com/google/error_analyzer/servlets/FulltextSearch.java:47]: make this a class variable.

File: src/main/java/com/google/error_analyzer/servlets/FulltextSearch.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+            String errorData = searchQuery.getErrorsAsString(indexFile, client);
```
Comment (nishantnahata) [inline src/main/java/com/google/error_analyzer/servlets/FulltextSearch.java:50]: Converting to json can be moved here in the servlet.

File: src/test/java/com/google/error_analyzer/FulltextSearchQueryTest.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+public final class FulltextSearchQueryTest{
+    private String indexFile = "trial_index"; //later fetched from request
+    RestHighLevelClient client;
+    @Before
+    public void setUp() {
+        client = new RestHighLevelClient(RestClient.builder(new HttpHost("35.194.181.238", 9200, "http")));
+    }
+
+@Test
+public void KeywordSearchQuery(){
```
Comment (nishantnahata) [inline src/test/java/com/google/error_analyzer/FulltextSearchQueryTest.java:42]: lower camelCase.

File: src/test/java/com/google/error_analyzer/FulltextSearchQueryTest.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+@RunWith(JUnit4.class)
+public final class FulltextSearchQueryTest{
+    private String indexFile = "trial_index"; //later fetched from request
+    RestHighLevelClient client;
+    @Before
+    public void setUp() {
+        client = new RestHighLevelClient(RestClient.builder(new HttpHost("35.194.181.238", 9200, "http")));
+    }
+
+@Test
```
Comment (nishantnahata) [inline src/test/java/com/google/error_analyzer/FulltextSearchQueryTest.java:41]: Also please maintain the indentation across the code.

File: src/test/java/com/google/error_analyzer/FulltextSearchQueryTest.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public final class FulltextSearchQueryTest{
+    private String indexFile = "trial_index"; //later fetched from request
+    RestHighLevelClient client;
+    @Before
+    public void setUp() {
+        client = new RestHighLevelClient(RestClient.builder(new HttpHost("35.194.181.238", 9200, "http")));
```
Comment (nishantnahata) [inline src/test/java/com/google/error_analyzer/FulltextSearchQueryTest.java:38]: Please remove the host IP.

Comment (nishantnahata) [reply src/test/java/com/google/error_analyzer/FulltextSearchQueryTest.java:38]: Also unit test shouldn't make a call to elasticsearch.

File: src/main/java/com/google/error_analyzer/backend/FulltextSearchQuery.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+* This class is for fulltext search query on a given index file using the keywords query string.
+* Can return Search hits as an Array of Errorline objects or json sring.
+* Depends on the RestHighLevelClientprovided and index filename provided in public functions.
+*/
```
[the remark this answers is not in the document]
Comment (Rythmag) [reply src/main/java/com/google/error_analyzer/backend/FulltextSearchQuery.java:19]: ok

File: src/main/java/com/google/error_analyzer/backend/FulltextSearchQuery.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        searchRequest.source(searchSourceBuilder); 
+
+        SearchResponse searchResponse = client.search(searchRequest, RequestOptions.DEFAULT);
+        SearchHits hits = searchResponse.getHits();
+        return hits;
+    }
+
+    // return log text from a hit.
+    private String extractLogText(SearchHit hit){
+        String sourceAsString = hit.getSourceAsString();
```
[the remark this answers is not in the document]
Comment (Rythmag) [reply src/main/java/com/google/error_analyzer/backend/FulltextSearchQuery.java:92]: It is used by getLogData to create Errorline object. I'll add the plugin

File: src/test/java/com/google/error_analyzer/FulltextSearchQueryTest.java (hunk 7 above)
Comment (Rythmag) [reply src/test/java/com/google/error_analyzer/FulltextSearchQueryTest.java:38]: We are working on the dao class.

File: src/main/java/com/google/error_analyzer/backend/FulltextSearchQuery.java [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+    }
+    
+    //Create Errorline array from search hits.
+    private ArrayList<ErrorLine> getLogData(SearchHits hits){
+
+        ArrayList<ErrorLine> errorMsgs = new ArrayList<>();
+        for (SearchHit hit : hits){
+            int lineNumber = extractLogLineNumber(hit);
+            if( lineNumber != -1){
+                errorMsgs.add(new ErrorLine( extractLogText(hit),lineNumber) );
```
Comment (nishantnahata) [inline src/main/java/com/google/error_analyzer/backend/FulltextSearchQuery.java:123]: Will there be a case when line numbers might repeat?

Comment (Rythmag) [reply src/main/java/com/google/error_analyzer/backend/FulltextSearchQuery.java]: We should face this issue. While storing logs, logLineNumber is also the id of the document

File: src/main/java/com/google/error_analyzer/servlets/FulltextSearch.java (hunk 4 above)
Comment (Rythmag) [reply src/main/java/com/google/error_analyzer/servlets/FulltextSearch.java]: Updated to move the json conversion to servlet.

File: src/main/java/com/google/error_analyzer/backend/FulltextSearchQuery.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (Rythmag) [reply src/main/java/com/google/error_analyzer/backend/FulltextSearchQuery.java]: Updated

**Comment by nishantnahata** [comment]
Please resolve conflicts.

**Comment by shudhatmajain** [comment]
Closing the PR.
This will still be available for reference in Closed PR tab.

**Pull request closed without merging**